### PR TITLE
Polymorphic descriptors

### PR DESCRIPTION
**Pull request by qwwdfsad**

### CODE REVIEW

File: runtime/commonMain/src/kotlinx/serialization/Polymorphic.kt [hunk 1]
```diff
@@ -75,5 +74,12 @@ public val PolymorphicClassDescriptor = SerialDescriptor("kotlinx.serialization.
  * @see SerializersModuleBuilder.polymorphic
  */
 public class PolymorphicSerializer<T : Any>(override val baseClass: KClass<T>) : AbstractPolymorphicSerializer<T>() {
-    public override val descriptor: SerialDescriptor = PolymorphicClassDescriptor
+    public override val descriptor: SerialDescriptor =
+        SerialDescriptor("kotlinx.serialization.Polymorphic", PolymorphicKind.OPEN) {
+            element("type", String.serializer().descriptor)
+            element(
+                "value",
+                SerialDescriptor("kotlinx.serialization.Polymorphic.${baseClass.simpleName}", UnionKind.CONTEXTUAL)
```
Comment (qwwdfsad) [inline runtime/commonMain/src/kotlinx/serialization/Polymorphic.kt]: Added class simple name for the SD FQN for the sake of observability and debuggability.
It doesn't really matter, just makes things more straightforward

File: formats/protobuf/commonTest/src/kotlinx/serialization/protobuf/ProtobufPolymorphismTest.kt [hunk 2]
```diff
@@ -24,4 +24,17 @@ class ProtobufPolymorphismTest {
...
+    sealed class Single {
+        @Serializable
+        class Impl : Single()
+    }
+
+    @Test
+    fun testSingleSealedClass() {
+        val expected =
+            "0a436b6f746c696e782e73657269616c697a6174696f6e2e70726f746f6275662e50726f746f627566506f6c796d6f72706869736d546573742e53696e676c652e496d706c1200"
+        assertEquals(expected, ProtoBuf.dumps(Single.serializer(), Single.Impl()))
```
Comment (sandwwraith) [inline formats/protobuf/commonTest/src/kotlinx/serialization/protobuf/ProtobufPolymorphismTest.kt]: Maybe check deserialization as well?

File: runtime/commonMain/src/kotlinx/serialization/Polymorphic.kt (hunk 1 above)
Comment (sandwwraith) [reply runtime/commonMain/src/kotlinx/serialization/Polymorphic.kt]: Maybe
```suggestion
                SerialDescriptor("kotlinx.serialization.Polymorphic<${baseClass.simpleName}>", UnionKind.CONTEXTUAL)
```
to be less cryptic?

File: runtime/commonMain/src/kotlinx/serialization/SealedSerializer.kt [hunk 3]
```diff
@@ -72,10 +73,16 @@ public class SealedClassSerializer<T : Any>(
 ) : AbstractPolymorphicSerializer<T>() {
 
     override val descriptor: SerialDescriptor = SerialDescriptor(serialName, PolymorphicKind.SEALED) {
-        subclassSerializers.forEach {
-            val d = it.descriptor
-            element(d.serialName, d)
-        }
+        element("type", String.serializer().descriptor)
+        val elementDescriptor =
+            SerialDescriptor("kotlinx.serialization.Sealed.${baseClass.simpleName}", PolymorphicKind.SEALED) {
```
Comment (sandwwraith) [inline runtime/commonMain/src/kotlinx/serialization/SealedSerializer.kt]:
```suggestion
            SerialDescriptor("kotlinx.serialization.Sealed<${baseClass.simpleName}>", PolymorphicKind.CONTEXTUAL) {
```

File: runtime/commonMain/src/kotlinx/serialization/SealedSerializer.kt [hunk 4]
```diff
@@ -84,7 +91,8 @@ public class SealedClassSerializer<T : Any>(
     init {
         require(subclasses.size == subclassSerializers.size) {
             "Arrays of classes and serializers must have the same length," +
-                    " got arrays: ${subclasses.contentToString()}, ${subclassSerializers.contentToString()}"
+                    " got arrays: ${subclasses.contentToString()}, ${subclassSerializers.contentToString()}\n" +
+                    "Please ensure that @Serializable annotation is present on each sealed subclass"
```
Comment (sandwwraith) [inline runtime/commonMain/src/kotlinx/serialization/SealedSerializer.kt]: Plugin automatically filters out non-serialziable subclasses, afair. So this error may happen only on manual construction of sealed serializer.

**Pull request merged**